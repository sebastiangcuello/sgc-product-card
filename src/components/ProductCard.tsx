import React, { CSSProperties, createContext } from 'react';
import { useProduct } from '../hooks/useProduct'
import { InitialValues, Product, ProductCardHandlers, ProductContextProps, onChangeArgs } from '../interfaces/interfaces';
import styles from '../styles/styles.module.css'

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;

export interface Props {
    product: Product;
    //children?: ReactElement | ReactElement[];
    children: ( args?: ProductCardHandlers ) => JSX.Element;
    className?: string;
    style?: CSSProperties;
    onChange?: ( args: onChangeArgs ) => void;
    value?: number;
    initialValues?: InitialValues;
}

export const ProductCard = ({ children, product, className, style, onChange, value, initialValues }: Props) => {
  
    const { counter, increaseBy, maxCount, isMaxCountReached, reset } = useProduct({ 
        onChange,
        product,
        value,
        initialValues
    })

    return (
        // es un componente de react que permite pasar datos a los componentes hijos
        <Provider value={{
            counter,
            increaseBy,
            maxCount,
            product
        }}> 
            <div 
                className={ `${ styles.productCard } ${ className } ` }
                style= { style }
            >
                { children({
                    count: counter,
                    isMaxCountReached,
                    maxCount: initialValues?.maxCount,
                    product,
                    increaseBy,
                    reset
                }) }
              </div>
        </Provider>

    )
}

// ProductCard.Title = ProductTitle;
// ProductCard.Image = ProductImage;
// ProductCard.Buttons = ProductButtons;