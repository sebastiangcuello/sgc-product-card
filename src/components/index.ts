import { ProductCard as ProductCardHOC } from "./ProductCard";
import { ProductCardHOCProps } from "../interfaces/interfaces";

import { ProductButtons } from "./ProductButtons";
import { ProductImage } from "./ProductImage";
import { ProductTitle } from "./ProductTitle";

export { ProductButtons } from "./ProductButtons";
export { ProductImage } from "./ProductImage";
export { ProductTitle } from "./ProductTitle";

//Con Object.assign podemos asignarle propiedades a un objeto
//En este caso le asignamos las propiedades Image, Title y Buttons
export const ProductCard: ProductCardHOCProps = Object.assign( ProductCardHOC, {
    Image: ProductImage,
    Title: ProductTitle,
    Buttons: ProductButtons,
});

export default ProductCard;