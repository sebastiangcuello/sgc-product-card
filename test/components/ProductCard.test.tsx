
import React from "react";
import renderer from 'react-test-renderer';
import { ProductCard } from "../../src/components";
import { product1 } from "../data/product";

describe('ProductCard', () => { 
    
    test('should show the component correctly', () => {

        const wrapper = renderer.create(
            <ProductCard 
                product={ product1 }
            >
                {
                    () => (
                        <h1>Product Card</h1>
                    )
                }
            </ProductCard>
        );

        expect( wrapper.toJSON() ).toMatchSnapshot();

    })

    //No funciona...
    // test('should increment the counter', () => { 
        
    //     const wrapper = renderer.create(
    //         <ProductCard 
    //             product={ product1 }
    //             initialValues={{
    //                 count: 0,
    //                 maxCount: 10,
    //               }}
    //         >
    //                 {
    //                     ( args ) => (
    //                         <>
    //                             <h1>Product Card</h1>
    //                             <span>{ args?.count }</span>
    //                             <button onClick={ () => args?.increaseBy(1) }></button>
    //                         </>
    //                     )
    //                 }
    //         </ProductCard>
    //     );

    //     let tree = wrapper.toJSON();
    //     expect( tree ).toMatchSnapshot();

    //     console.log(tree);

    //     act(() => {
    //         (tree as any).children[2].props.onClick();
    //     })

    //     console.log(tree);

    //     tree = wrapper.toJSON();
    //     expect( (tree as any).children[1].children[0] ).toBe('1');
    //  })
 })

 