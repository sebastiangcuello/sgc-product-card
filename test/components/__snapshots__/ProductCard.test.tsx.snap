// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard should show the component correctly 1`] = `
<div
  className="productCard undefined "
>
  <h1>
    Product Card
  </h1>
</div>
`;
